Fix display LogInComponent turning blank when entering wrong username/password combination

alexandrevryghem opened the pull request:
## References
* Fixes #2515
* Related #2414

## Description
Fixes bug where the `LogInComponent` turns blank when submitting invalid username/password combination.

## Instructions for Reviewers
Instead of hiding the `LogInContainerComponent` by checking it's content, now it hides it by checking if the login method has a component with a `rendersAuthMethodType` decorator.

**Guidance for how to test or review this PR:**
You can test it locally but I wasn't able to reproduce the bug myself with my local setup

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

alanorth commented:
Reviewed by testing on DSpace 7.6.

I haven't managed to pinpoint the circumstances where it occurs, but I get a blank login menu very frequently when running in dev mode. With this patch I have not noticed the blank menu even once, and everything is working as normal as far as I can tell. I'm +1 by testing. Haven't reviewed the code yet.

vins01-4science commented on src/app/shared/log-in/container/log-in-container.component.ts:
```diff
@@ -51,8 +53,12 @@
-  getAuthMethodContent(): string {
-      return rendersAuthMethodType(this.authMethod.authMethodType);
+  getAuthMethodContent(): Component {
+    const component = rendersAuthMethodType(this.authMethod.authMethodType);
+    if (component === undefined) {
+      this.elRef.nativeElement.classList.add('d-none');
+    }
+    return component;
```
If the problem is that the `component` is undefined, it would be better to filter out all the `authMethodType` that are not configured. To achieve this behavior, you should filter them inside the `log-in-component` and add the related method inside the `login-in.methods-decorator`.
Do you agree @alexandrevryghem?

alexandrevryghem replied:
Thnx for the review @vins01-4science, I indeed do agree that it would be even better to filter them out earlier so that it never even generates a `LogInContainerComponent` for `AuthMethod`s without a `Component`. But I would prefer to just check if the `Component` is undefined though, so that it works dynamically, independently of which (future) `AuthMethod`s are added.

vins01-4science commented on src/app/shared/log-in/methods/log-in.methods-decorator.ts:
```diff
@@ -1,6 +1,6 @@
 import { AuthMethodType } from '../../../core/auth/models/auth.method-type';
 
-const authMethodsMap = new Map();
+const authMethodsMap: Map<AuthMethodType, any> = new Map();
```
Don't use `any`, it would be better to use `Component`

vins01-4science commented on src/app/shared/log-in/methods/log-in.methods-decorator.ts:
```diff
@@ -1,6 +1,6 @@
 
-const authMethodsMap = new Map();
+const authMethodsMap: Map<AuthMethodType, any> = new Map();
 
```
Here you should add the `hasAuthMethodType` function that will return `true` if the map has that target key configured

vins01-4science reviewed (changes requested):
Hello @alexandrevryghem ,

thank you for this PR, I've tested its behavior and it seems to fix that issue.
I'm glad you fixed it, but I think that we can improve the code by not using the `ElementRef` and the `d-none` class. Instead we should filter out for all the `AuthMethodType` not configured by extending the filter
```javascript
      map((methods: AuthMethod[]) => methods.filter((authMethod: AuthMethod) => authMethod.authMethodType !== AuthMethodType.Ip)),
```
inside the `log-in.component.ts` by using the new method inside the `log-in.methods-decorator.ts`.

I would like to ask for these changes, Do you agree with these requests?

Thank you.

alexandrevryghem commented:
@vins01-4science, I implemented your feedback, and I also added a small commit to fix an error that I found while making the updates.
When you are running Angular, but your backend is down, the error `ERROR TypeError: authorizations.filter is not a function` is logged. This error occurs because `AuthorizationDataService#isAuthorized` returns `false` when an error is thrown, and `oneAuthorizationMatchesFeature` tries to iterate over it, which is obviously not possible. To fix this, an empty array needs to be returned, and this will still result in an `observableOf(false)` being returned by `AuthorizationDataService#isAuthorized`.

vins01-4science reviewed (approved):
Thank you @alexandrevryghem for the changes, everything seems to be correct to me :rocket: .
I am :+1:  on this one.

tdonohue reviewed (approved):
👍 Thanks @alexandrevryghem !  Verified this fixes the bug & that reordering of login options (Shib, Password, etc) still works properly

tdonohue commented:
Ported to 7.6.x manually via #2575
